sentinel-core/src/main/java/com/alibaba/csp/sentinel/slots/automatic/AutomaticRuleManager.java

```diff
@@ -0,0 +1,316 @@
+package com.alibaba.csp.sentinel.slots.automatic;
+
+import com.alibaba.csp.sentinel.context.Context;
+import com.alibaba.csp.sentinel.node.DefaultNode;
+import com.alibaba.csp.sentinel.node.Node;
+import com.alibaba.csp.sentinel.slotchain.ResourceWrapper;
+import com.alibaba.csp.sentinel.slots.block.BlockException;
+import com.alibaba.csp.sentinel.slots.block.RuleConstant;
+import com.alibaba.csp.sentinel.slots.block.flow.FlowException;
+import com.alibaba.csp.sentinel.slots.block.flow.FlowRule;
+import com.alibaba.csp.sentinel.slots.block.flow.controller.DefaultController;
+import com.alibaba.csp.sentinel.slots.system.SystemStatusListener;
+
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * AutomaticRuleManager 维护一套监控数据记录和流控规则
+ * 流控规则根据实时流量动态计算
+ *
+ * @author Li Yudong
+ */
+public class AutomaticRuleManager {
+
+    /**
+     * 流控规则
+     */
+    private static Map<String, FlowRule> rules = new ConcurrentHashMap<String, FlowRule>();
+
+    /**
+     * 监控数据
+     */
+    private static Map<String, Integer> qpsRecord = new ConcurrentHashMap<String, Integer>();
```

Comment:
Can we abstract a universal data structure for adaptive statistics for each resource?